### DISCUSSION TITLE
`sourceSet` Feature for Image

### DISCUSSION THREAD

**Discussion by yuliakbard** [mecha-cms/mecha#178] [Ideas]
Saya pengguna blogger yang masih dalam pembelajaran mengenai PHP, ingin transisi ke Mecha. saya mengharapkan fitur sourceSet untuk gambar. sesuatu yang mirip `sourceSet(data:post.featuredImage, [240,330,600])`

saya membayangkan cara menggunakannya, tinggal menuliskan seperti ini `<?=self::sourceSet($image, [240,330,600]);?>`

Dengan Output:
`/gambar1-w240.jpg 240w, /gambar1-w330.jpg 330w, /gambar1-w600.jpg 600w`
atau
`/w240/gambar1.jpg 240w, /w330/gambar1.jpg 330w, /w600/gambar1.jpg 600w`

Dengan banyak fitur-fitur yang mirip seperti blogger, dengan ini akan membuat para blogger pemula yang ingin lebih advance dalam mengelola Website akan migrasi menggunakan Mecha CMS, begitu juga para designer template blogger yang akan membuat desain template untuk Mecha yang lebih kreatif

**Comment by taufik-nurrohman** [comment]
Ide kamu sebenarnya masuk mas, cuma akan lebih efektif kalau pakai fungsi. Mas bisa buat fungsi `source_set()` misalnya, kemudian mengubah URL gambar yang ada di dalamnya menjadi thumbnail menggunakan ekstensi [`image`](https://github.com/mecha-cms/x.image):

~~~ php
function source_set(string $source, array $set = []) {
    $image = new Image($source);
    // Resize, crop, dsb…
}
~~~

Kendalanya ada di cara mengubah gambar tersebut menjadi URL yang baru. Kita bisa pakai rute tapi akan butuh langkah tambahan lagi. Cara yang paling mudah ya dengan menggunakan layanan pihak ke tiga, misalnya [Statically](https://statically.io):

~~~ php
function source_set(string $source, array $set = []) {
    $out = [];
    $source = ltrim($source, '/');
    // <https://statically.io/docs/using-images>
    foreach ($set as $v) {
        $out[] = 'https://cdn.statically.io/img/' . $_SERVER['HTTP_HOST'] . '/w=' . $v . '/' . $source . ' ' . $v . 'w';
    }
    return implode(', ', $out);
}
~~~

Contoh penggunaan:

~~~ php
echo source_set('/lot/asset/test.jpg', [100, 200, 300]);
~~~

**Reply by yuliakbard** [reply]
kang, ini output srcsetnya seperti ini, gak muncul gambarnya
`https://cdn.statically.io/img/www.domain.com/w=100/https://www.domain.com/lot/asset/.cache/26/f9/05/e4/28/52/64/9d/e8/d6/b1/05/78/58/2b/5e.jpg 100w, https://cdn.statically.io/img/www.domain.com/w=200/https://www.domain.com/lot/asset/.cache/26/f9/05/e4/28/52/64/9d/e8/d6/b1/05/78/58/2b/5e.jpg 200w, https://cdn.statically.io/img/www.domain.com/w=300/https://www.domain.com/lot/asset/.cache/26/f9/05/e4/28/52/64/9d/e8/d6/b1/05/78/58/2b/5e.jpg 300w`

pas ane kroscek secara manual url gambar yang muncul itu seperti ini:
`https://cdn.statically.io/img/www.domain.com/w=300/lot/asset/.cache/26/f9/05/e4/28/52/64/9d/e8/d6/b1/05/78/58/2b/5e.jpg` 

Gimana cara mengatasinya kang?

ane panggilnya seperti ini
`<?=source_set($page->image(600,600,100), [100, 200, 300]);?>`

**Reply by taufik-nurrohman** [reply]
### < v3.0.0

~~~ php
<?= source_set(URL::short($page->image), [100, 200, 300]); ?>
~~~

### >= v3.0.0

~~~ php
<?= source_set(short($page->image), [100, 200, 300]); ?>
~~~

**Reply by yuliakbard** [reply]
mecha v3.0.0 itu udah versi stabil kah kang?

**Reply by taufik-nurrohman** [reply]
Belum mas 😄

Targetnya nanti mau Saya lengkapi dengan fitur basis data (sebagai ekstensi terpisah).

Cek proyek eksperimen basis data Saya [di sini](https://github.com/taufik-nurrohman/table-manager).

**Reply by yuliakbard** [reply]
pantesan, saya coba jalankan di localhost muncul pesan error. hehehe. oke deh kang, nanti di pantau deh update-an nya

**Comment by lokersumatera** [comment]
cara menggunakannya dalam bentuk file php gimana penulisan functionnya kang?
```
function source_set(string $source, array $set = []) {
    $out = [];
    $source = ltrim($source, '/');
    // <https://statically.io/docs/using-images>
    foreach ($set as $v) {
        $out[] = 'https://cdn.statically.io/img/' . $_SERVER['HTTP_HOST'] . '/w=' . $v . '/' . $source . ' ' . $v . 'w';
    }
    return implode(', ', $out);
}
```

biar bisa dipanggil seperti ini:
`<?=self::source_set(URL::short($page->image), [240, 360, 600])?>`

**Reply by taufik-nurrohman** [reply]
Tambahin metode aja ke kelas `Layout`:

~~~ php
function source_set(string $source, array $set = []) {}

Layout::_('source_set', 'source_set');

// Contoh penggunaan
echo Layout::source_set('/path/to/image.jpg', [200, 300]);

// Atau pakai `self` saja kalau di dalam berkas layout
echo self::source_set('/path/to/image.jpg', [200, 300]);
~~~

**Reply by yuliakbard** [reply]
kalau misalnya size yang mau ditampilkan `$set` hanya 1, bagaimana kondisional nya seperti apa kang?

`echo self::source_set('/path/to/image.jpg', [200]);`

saya pake cara seperti ini, hanya url domain saja yang tampil? 

```
function source_set(string $source, array $set = []) {
    $out = [];
    $source = ltrim($source, '/');
if(count($set) > 1){
    foreach ($set as $v) {
        $out[] = 'https://cdn.statically.io/img/' . $_SERVER['HTTP_HOST'] . '/w=' . $v . '/' . $source . ' ' . $v . 'w';
    }
}else{
$out[] = 'https://cdn.statically.io/img/' . $_SERVER['HTTP_HOST'] . '/w=' . $v . '/' . $source;
}
    return implode(', ', $out);
}
```

**Reply by taufik-nurrohman** [reply]
Seharusnya tidak perlu mengedit kode sudah bisa dipakai saja.

**Reply by yuliakbard** [reply]
`$out[] = 'https://cdn.statically.io/img/' . $_SERVER['HTTP_HOST'] . '/w=' . $v . '/' . $source . ' ' . $v . 'w';`


`https://2.bp.blogspot.com/-nr_EHIP99Uk/Xva13D69z2I/AAAAAAAAEf4/Erx0e0zDusA3cIEdYDda4V__EwPJ5tzeQCPcBGAYYCw/s600-rw/lokersumatera-Grand%2BSari%2BHotel.jpg`

hasilnya jadi `https://cdn.statically.io/img/lokersumatera.my.id/w=600/https://2.bp.blogspot.com/-nr_EHIP99Uk/Xva13D69z2I/AAAAAAAAEf4/Erx0e0zDusA3cIEdYDda4V__EwPJ5tzeQCPcBGAYYCw/s600-rw/lokersumatera-Grand%2BSari%2BHotel.jpg`

apa url gambar nya di trim dulu kah, atau diapakan dulu ya kang?

kalau saya pakai extension statically, gak bisa ngerubah url gambar yang di srcset

**Reply by taufik-nurrohman** [reply]
Agak susah karena udah beda domain.

**Reply by yuliakbard** [reply]
ada cara kondisional gak kang?

yg satu pake ini

`$out[] = 'https://cdn.statically.io/img/' . $_SERVER['HTTP_HOST'] . '/w=' . $v . '/' . $source . ' ' . $v . 'w';`

untuk yang dari blogger pake gini

`$out[] = 'https://cdn.statically.io/img/'  . $source . '?w=' . $v .' ' . $v . 'w';`

url eksternal gak bisa di trim atau ambil host/ path nya aja ya kang?

atau ada alternatif lain gak kang?

**Reply by taufik-nurrohman** [reply]
~~~ php
foreach ($set as $v) {
    if (false !== strpos($v, '://')) {
        // Eksternal (absolut)
    } else {
        // Internal (relatif)
    }
}
~~~

**Reply by yuliakbard** [reply]
kang, kalau url gambar dari blogger, gimana replace size nya kang, ane masih mau ngandelin penyimpanan gambar di blogger. gak kuat dana kalau nyimpan di hosting. hehehe

```
$url = 'https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEj98Uumck9jAwE001VwF5fSs7NZ0pvWTNz51JlZJ7JetW63sM0PdKkdV8THZz3zcq2HXATLLo8EH-9i6UcTqvs78f_EtWCUWI9QxDl1sJq6tRf8NpdqK8jrnkbuXTXQb2NlsGL_itKRaYCztSlmsFr6TtfdtB5XzX7P9L6r1OfVn7qGvvEgZB5ycCgv/s1280/Loker%20Sumatera%20-%20Barbekyu%20Steakhouse.jpg'
$replace = '/\/s[0-9]{2}-(w[0-9]+-)?c/'
$size = 's600'
```

**Reply by yuliakbard** [reply]
kalau misalkan, kita make url gambar secara eksternal, gimana cara nambahin gambar tersebut menjadi gambar featured image, yang bisa muncul juga di feed.xml? apa itu memungkinkan kang?

**Reply by taufik-nurrohman** [reply]
~~~ php
$url = preg_replace($replace, $size, $url);
~~~

**Reply by taufik-nurrohman** [reply]
Kalau pakai ekstensi `image` si seharusnya akan secara otomatis bisa.

**Reply by yuliakbard** [reply]
Kang, kenapa srcset ane gak jalan di gambar pertama di setiap halaman, gambar 2, 3, dst. ke render?

**Comment by yuliakbard** [comment]
Kang, kenapa disetiap hasil url sourceset pertama ane, outputnya NULL? di gambar ke dua, ke tiga outputnya normal

**Reply by taufik-nurrohman** [reply]
Coba kasih sampel fungsi-nya, dan isi laman-nya.

**Reply by yuliakbard** [reply]
ini fungsinya ane simpan di `source_set.php`

```
<?php
function source_set(string $source, array $set = []) {
    $out = [];
    $pattern = '/\/s[0-9]+/';  
foreach ($set as $v) {
    $replace = '/w' . $v.'-v2-rw-e7';
    $blogger_url = preg_replace($pattern, $replace, $source);
$out[] = strrchr($source,'://') !== false ? $blogger_url. '  '.$v .'w' : 'https://cdn.statically.io/img/'.$_SERVER['HTTP_HOST'].'/w='.$v.',f=auto'.$source.' '.$v.'w';
}   
    return implode(', ', $out);
}
self::_('source_set', 'source_set');
```

ini untuk penggunaan gambarnya disimpan di `thumbnail.php`

```
<div class="postThumb placeholder">
<div class='preloader'><div class='sp sp-wave'></div></div>
<?php if ($images = $page->images): ?>
  <?php if (count($images) > 1): ?>
<?php if(!$site->is('mobile')):?>
    <!-- Images as carousel -->
    <div class="slider">
    <?php foreach ($images as $image): ?>
    <div class="item-slider placeholder"><div class='preloader'><div class='sp sp-wave'></div></div><img class="f snippet-thumbnail w100 lazy" width="9" height="16" alt="<?= $page->title ?>" data-src="<?=$page->image(360, 360, 100)?>" data-srcset="<?=self::source_set(URL::short($page->image), [36, 180, 240, 360, 600]);?>"></div>
    <?php endforeach; ?>
    </div>
<?php else: ?>
    <amp-carousel class="slider"
      width="1"
  height="1"
  layout="responsive"
  type="slides"
  role="region"
  aria-label="Carousel with slide previews"
    >
<?php foreach ($images as $image): ?>
<div class="item-slider placeholder"><img class="f snippet-thumbnail w100 lazy" width="1" height="1" alt="<?= $page->title ?>" src="<?= $page->image(360,360, 100); ?>"></div>
<?php endforeach; ?>
</amp-carousel>
<?php endif;?>
<?php else: ?>
    <!-- Image as figure -->
<img class="f snippet-thumbnail w100 lazy" width="1" height="1" alt="<?= $page->title ?>" data-src="<?=reset($images)??$page->organization_logo; ?>" src="data:image/png;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=" srcset="<?=self::source_set(URL::short(reset($images))??$page->organization_logo, [36, 180, 240, 360, 600]);?>">
<?php endif; ?>
<?php else: ?>
  <?php if ($image = $page->image(360,360,100) ?? $page->organization_logo): ?>
<img class="f snippet-thumbnail w100 lazy single" width="1" height="1" alt="<?= $page->title ?>" data-src="<?= $image; ?>" src="data:image/png;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs="
data-srcset="<?=self::source_set(URL::short($image)??$page->organization_logo, [36, 180, 240, 360, 600]);?>">
  <?php else: ?>
    <!-- No image -->
    <img class="f snippet-thumbnail w100 lazy" width="1" height="1" alt="<?= $page->title ?>" data-src="/lot/asset/jpg/loker-sumatera--instagram.jpg" src="data:image/png;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs="
    data-srcset="<?=self::source_set("/lot/asset/jpg/loker-sumatera--instagram.jpg", [36, 180, 240, 360, 600]);?>">
<?php endif; ?>
<?php endif; ?>
</div>
```

ini bagian dari `list.php` yang akan menampilkan article di `pages.php`

```
<div class="list">
    <div class="forImage">
            <?=self::thumbnail(['page' => $page]);?>
    </div>
    <div class="forSummary">
        <div class="forJobTitle">
            <h2 class="post-title">
<a href="<?= $page->link ? $page->link : $page->url ?>"><?=$page->job_position ? $page->job_position : $page->title; ?></a>
            </h2>
        </div>
        <div clss="forCompany">
            <?=$page->organization_name;?>
        </div>
        <div class="forView">
            <?= self::view(['page' => $page]) ?>
        </div>
            <span class="forTimeAgo">
                <?= self::timeago(['page' => $page]); ?>
            </span>
    </div>
</div>
```

ini isi `pages.php`

```
<?= self::before() ?>
<main class="widthMain">
    <article id="page:<?= $page->id ?>" class="<?=$page->name;?>">
   <header>
    <div class="widget-title">
    <?php if(!$site->is('home')&&$site->has('pages')): ?>
    <h2><?= $page->title ?> (<?=$pages->count;?>)</h2>
    <?php else: ?>
    <h2>Lowongan Kerja Terbaru</h2>
    <?php endif;?>
    </div>
    </header>
    <div class="blog-posts">
            <?php foreach ($pages as $page) : ?>
            <?= self::list(['page'=> $page]) ?>
        <?php endforeach; ?>
<?= self::pager(); ?>   
<?= self::pager('steps'); ?>       
         </div>
    </article>
    <aside>
<?= self::popular(['page'=> $page]) ?>
    </aside>
</main>
<?= self::after() ?>
```

link: `lokersumatera.my.id`

**Reply by yuliakbard** [reply]
hasil null nya sendiri ane test pake `var_dump()`; karena bingung cari error nya dimana. hehehe

**Reply by taufik-nurrohman** [reply]
Naruhnya jangan di `source_set.php` mas tapi di `index.php`. Sekali register aja cukup. Terus bagian `self::` ganti jadi `Layout`. Itu fail `source_set.php` dihapus aja, udah digantiin sama `Layout::_('source_set')`

**Reply by yuliakbard** [reply]
solved kang

**Reply by yuliakbard** [reply]
tinggal beberapa permasalahan basiclagi nih, baru bisa dengan tenang ane desain website menggunakan mecha.

1.  sourceset [done]
2.  next pager [not solved] untuk versi 2.6.4 alternatifnya baru menggunakan javascript.
3.  url gambar eksternal gak kebaca di feed.xml [not solved]
4.  custom panel, masih progress. karena belum ada dokumentasi lengkap untuk penggunaannya [on progress]
5.  menyisipkan iklan di setiap beberapa artikel di halaman index/ homepage [not solved]
6.  untuk tag label, hanya bisa linear. kalau ada sub halaman, tag label nya ngikut di sub halaman, mau ane walaupun ada sub halaman tagnya itu tetap ngikut parentnya.

**Reply by taufik-nurrohman** [reply]
Untuk poin 2, pakai versi pre-release harusnya sudah bisa. Di repo dan situs web memang belum tampak, tapi bisa dicoba pasang pakai Composer.

~~~ sh
composer create-project mecha-cms/mecha .
composer require mecha-cms/x.panel
~~~

Poin 3 bisa dijadikan isu untuk ekstensi `feed` dan `image`.

Poin 5 dan 6 bisa dibuat sebagai diskusi tersendiri.

**Reply by yuliakbard** [reply]
penggunaan composer bagus juga nih kang, dari kemarin ane cuma download langsung di github. wkwk

**Reply by yuliakbard** [reply]
cara menginstal layoutnya disimpan di folder y kah kang?

**Reply by taufik-nurrohman** [reply]
Iya. Versi 3.0.0 nanti akan mengembalikan fitur _layout switcher_.

**Comment by yuliakbard** [comment]
untuk url gambar eksternal yang tidak terbaca di feed.xml gimana cara mengatasinya kang? sedangkan gambar internal bisa muncul di feed.xml

**Reply by yuliakbard** [reply]
untuk panelnya ane menggunakan

```
        $_['lot']['desk']['lot']['form']['lot'][1]['lot']['tabs']['lot']['page']['lot']['fields']['lot']['organization_logo'] = [
            'title' => 'Logo Perusahaan',
            'description' => 'isi Link Logo.',
            'type' => 'field/link',
            'name' => 'page[organization_logo]',
            'value' => $page['organization_logo'],
            'width' => true,
            'stack' => 2.8
        ];
```
penggunaan, yang ane terapkan di layout seperti ini:  `$image = $page->organization_logo??$page->image`
